core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -202,6 +203,12 @@ public int unfinishedRequests() {
 
     @Override
     public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        if (responseEncoder != null) {
+            // Immediately close responseEncoder so that a late response is completed with
+            // a ClosedSessionException.
+            responseEncoder.close();
+        }
```

Comment:
Sorry, I'm having a little trouble understanding this change 😅 

Is it still possible that `write` succeeds but `channelInactive` is invoked before `WriteFutureListener` is invoked? Or is the following comment not relevant anymore? 🤔 

https://github.com/line/armeria/pull/4508/files#diff-8c8fd516b9814fa10a52ccad8660a3383cc0862c1e0cc6f91b0a456eb11a3641R212-R221

Reply:
I guess the delay of 1 second is used to complete a request and its response.
So closing `responseEncoder.close()` early does not break the original workaround.

Reply:
Ah I misunderstood the intention of this change - I understood this change was made to avoid writing to a closed channel (and exposing `StacklessClosedChannelException`).
Sounds good to me 👍